pkg/entities/ie.go

```diff
@@ -390,7 +505,154 @@ func EncodeToIEDataType(dataType IEDataType, val interface{}) ([]byte, error) {
+	case Signed32:
+		v, err := element.GetSigned32Value()
+		if err != nil {
+			return err
+		}
+		binary.BigEndian.PutUint32(buffer[index:], uint32(v))
+	case Signed64:
+		v, err := element.GetSigned64Value()
+		if err != nil {
+			return err
+		}
+		binary.BigEndian.PutUint64(buffer[index:], uint64(v))
+	case Float32:
+		v, err := element.GetFloat32Value()
+		if err != nil {
+			return err
+		}
+		binary.BigEndian.PutUint32(buffer[index:], math.Float32bits(v))
+	case Float64:
+		v, err := element.GetFloat64Value()
+		if err != nil {
+			return err
+		}
+		binary.BigEndian.PutUint64(buffer[index:], math.Float64bits(v))
+	case Boolean:
+		v, err := element.GetBooleanValue()
+		if err != nil {
+			return err
+		}
+		// Following boolean spec from RFC7011
+		indicator := byte(int8(1))
+		if !v {
+			indicator = byte(int8(2))
+		}
+		copy(buffer[index:index+1], []byte{indicator})
+	case DateTimeSeconds:
+		v, err := element.GetUnsigned32Value()
+		if err != nil {
+			return err
+		}
+		binary.BigEndian.PutUint32(buffer[index:], v)
+	case DateTimeMilliseconds:
+		v, err := element.GetUnsigned64Value()
+		if err != nil {
+			return err
+		}
+		binary.BigEndian.PutUint64(buffer[index:], v)
+		// Currently only supporting seconds and milliseconds
+	case DateTimeMicroseconds, DateTimeNanoseconds:
+		// TODO: RFC 7011 has extra spec for these data types. Need to follow that
+		return fmt.Errorf("API does not support micro and nano seconds types yet")
+	case MacAddress:
+		// Expects net.Hardware type
+		v, err := element.GetMacAddressValue()
+		if err != nil {
+			return err
+		}
+		copy(buffer[index:], v)
+	case Ipv4Address:
+		// Expects net.IP type
+		v, err := element.GetIPAddressValue()
+		if err != nil {
+			return err
+		}
+		if ipv4Add := v.To4(); ipv4Add != nil {
+			copy(buffer[index:], ipv4Add)
+		} else {
+			return fmt.Errorf("provided IP %v does not belong to IPv4 address family", v)
+		}
+	case Ipv6Address:
+		// Expects net.IP type
+		v, err := element.GetIPAddressValue()
+		if err != nil {
+			return err
+		}
+		if ipv6Add := v.To16(); ipv6Add != nil {
+			copy(buffer[index:], ipv6Add)
+		} else {
+			return fmt.Errorf("provided IPv6 address %v is not of correct length", v)
+		}
+	case String:
+		v, err := element.GetStringValue()
+		if err != nil {
+			return err
+		}
+		if len(v) < 255 {
+			buffer[index] = uint8(len(v))
+			for i, b := range v {
+				buffer[i+index+1] = byte(b)
+			}
+		} else if len(v) < 65535 {
+			buffer[index] = byte(255)
+			binary.BigEndian.PutUint16(buffer[index+1:index+3], uint16(len(v)))
+			for i, b := range v {
+				buffer[i+index+3] = byte(b)
+			}
+		}
+	default:
+		return fmt.Errorf("API supports only valid information elements with datatypes given in RFC7011")
+	}
+	return nil
+}
+
+// encodeToBuff is to encode data to specific type to the buff. This is only used
+// for testing.
 func encodeToBuff(dataType IEDataType, val interface{}, length int, buffer []byte, index int) error {
```

Comment:
Is it possible to remove either `encodeToBuff` or `EncodeToIEDataType`? I feel like we have redundant encoding methods for testing purpose.

Reply:
Removed `encodeToBuff` as it was used only for tests now and not needed anywhere.